main.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+module Enumerable # rubocop:disable Metrics/ModuleLength
+  def my_each
+    return to_enum unless block_given?
+
+    array = is_a?(Range) ? to_a : self
+    i = 0
+    while i < array.length
+      yield(array[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    array = is_a?(Range) ? to_a : self
+    i = 0
+    while i < array.length
+      yield(array[i], i)
+      i += 1
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    new_array = []
+    my_each { |x| new_array.push(x) if yield(x) }
+    new_array
+  end
+
+  def my_all?(pattern = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+    if block_given?
+      my_each { |x| return false unless yield(x) }
+    elsif pattern.class == Class
+      my_each { |x| return false unless x.is_a? pattern }
+    elsif pattern.class == Regexp
+      my_each { |x| return false if pattern.match? x }
+    else
+      my_each { |x| return false if x == pattern }
+    end
+    true
+  end
```

Comment:
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern